### PR TITLE
rework robotupload, add tests

### PR DESCRIPTION
**Pull request by nyirit**

### CODE REVIEW

File: scoap3/modules/robotupload/util.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    _add_additional_info(obj)
+
+    return obj
+
+
+def _add_additional_info(obj):
+    # map journal title and save new value.
+    journal_title = obj['publication_info'][0]['journal_title']
+    journal_title = JOURNAL_TITLE_MAPPING.get(journal_title, journal_title)
+    obj['publication_info'][0]['journal_title'] = journal_title
```
Comment (ammirate) [inline scoap3/modules/robotupload/util.py:48]: Do we expect this field to be always present?

File: scoap3/modules/robotupload/util.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    # get publisher based on journal title
+    publisher = JOURNAL_PUBLISHER_MAPPING.get(journal_title)
+
+    # add publisher if not exists
+    if 'imprints' in obj and 'publisher' not in obj['imprints'][0]:
+        obj['imprints'][0]['publisher'] = publisher
```
Comment (ammirate) [inline scoap3/modules/robotupload/util.py:55]: Same

File: scoap3/modules/robotupload/util.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    # add creation date if not exists
+    if 'record_creation_date' not in obj:
+        obj['record_creation_date'] = datetime.now().isoformat()
+
+    # add acquisition_source if not exists
+    if 'acquisition_source' not in obj:
```
Comment (ammirate) [inline scoap3/modules/robotupload/util.py:62]: This will never happen with the new version of `inspire-crawler` :D

File: scoap3/modules/robotupload/util.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            'date': datetime.now().isoformat()
+        }
+
+    # add material if not exists
+    if 'material' not in obj['publication_info'][0] and 'document_type' in obj:
+        obj['publication_info'][0]['material'] = obj['document_type'][0]
+
+    # remove document type, it should be in publication_info as material
+    if 'document_type' in obj:
+        del obj['document_type']
```
Comment (ammirate) [inline scoap3/modules/robotupload/util.py:75]: Why?

File: scoap3/modules/robotupload/views.py [hunk 5]
```diff
@@ -33,76 +35,63 @@ def handle_invalid_usage(error):
 
 
 def allowed_file(filename):
-    return '.' in filename and \
-           filename.rsplit('.', 1)[1].lower() in config.ROBOTUPLOAD_ALLOWED_EXTENSIONS
+    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ROBOTUPLOAD_ALLOWED_EXTENSIONS
```
Comment (ammirate) [inline scoap3/modules/robotupload/views.py]: This is not really the flask way of using config variables. If you read such a value in the following way, you don't need to redeploy the application if you want to change only the configuration :)
```python
from flask import current_app
my_val = current_app.config.get('ROBOTUPLOAD_ALLOWED_EXTENSIONS')
. . .
```

File: scoap3/modules/robotupload/views.py [hunk 6]
```diff
@@ -33,76 +35,63 @@ def handle_invalid_usage(error):
...
 
 
-@blueprint.route('/robotupload', methods=['POST'])
-def robotupload():
-    mode = None
+def _validate_request():
+    """Check if the request comes from a trusted source and parameters are valid."""
+    remote_addr = request.environ['REMOTE_ADDR']
+    if remote_addr not in ROBOTUPLOAD_ALLOWED_USERS:
```
Comment (ammirate) [inline scoap3/modules/robotupload/views.py]: same as above

File: scoap3/modules/robotupload/views.py [hunk 7]
```diff
@@ -33,76 +35,63 @@ def handle_invalid_usage(error):
...
 
-    if request.environ['REMOTE_ADDR'] not in config.ROBOTUPLOAD_ALLOWED_USERS:
-        raise InvalidUsage("Sorry, client IP %s cannot use the service." % request.environ['REMOTE_ADDR'],
-                           status_code=403)
     if 'file' not in request.files:
+        logger.warning('Robotupload accessed without a file specified. Remote addr: %s' % remote_addr)
         raise InvalidUsage("Please specify file body to input.")
+
     file = request.files['file']
     if file.filename == '':
```
Comment (ammirate) [inline scoap3/modules/robotupload/views.py]: I know is not part of this PR, but this is super ugly in Python... what about changing it to this?
```python
if not file.filename:
    . . .
```

File: scoap3/modules/robotupload/views.py [hunk 8]
```diff
@@ -33,76 +35,63 @@ def handle_invalid_usage(error):
...
-        raise InvalidUsage("Please specify file body to input.")
-    if file and allowed_file(file.filename):
-        filename = secure_filename(file.filename)
-        file.save(os.path.join(UPLOAD_FOLDER, filename))
-    else:
+        logger.warning('Robotupload accessed without a filename specified. Remote addr: %s' % remote_addr)
+        raise InvalidUsage("Please specify file body to input. Filename missing.")
+
+    if not file or not allowed_file(file.filename):
```
Comment (ammirate) [inline scoap3/modules/robotupload/views.py]: This check should happen before the previous one: if `file` is `None`, the previous check on `file.filename` will generate an exception.

File: scoap3/modules/robotupload/views.py [hunk 9]
```diff
@@ -33,76 +35,63 @@ def handle_invalid_usage(error):
...
-        CELERY_TASK_SERIALIZER='json',
-        CELERY_RESULT_SERIALIZER='json',
-    ))
-    celery.send_task(
-        API_ENDPOINT_DEFAULT,
-        kwargs={'results_data': obj}
-    )
 
+def _check_permission_for_journal(journal_title, remote_addr, package_name):
+    # check if the user has the right to upload an article with the provided journal title
```
Comment (ammirate) [inline scoap3/modules/robotupload/views.py]: Since this (and also `_validate_request`) are very critical functions, I would:
* remove the `_` in front
* replace the comment with a proper docstring
* write a bunch of unit tests for each function and every combination of the input, to be SURE that this check is doing what you expect :bomb:. Example: what if `journal_title` is an empty string? What if `package_name` is empty too? Does `package_name` need to be in a special format? ecc..

File: scoap3/modules/robotupload/views.py [hunk 10]
```diff
@@ -33,76 +35,63 @@ def handle_invalid_usage(error):
...
+        logger.warning('Wrong journal name in metadata for package: %s. Remote addr: %s' % (package_name, remote_addr))
+        raise InvalidUsage("Cannot submit such a file from this IP. (Wrong journal)")
+
+
+def handle_upload_request(apply_async=True):
+    """Handle articles that are pushed from publishers."""
+    _validate_request()
+
+    remote_addr = request.environ['REMOTE_ADDR']
+    file = request.files['file']
```
Comment (ammirate) [inline scoap3/modules/robotupload/views.py]: Better use a different name, `file` is a Python keyword :P

File: scoap3/modules/workflows/workflows/articles_upload.py [hunk 11]
```diff
@@ -116,13 +118,18 @@ def is_record_in_db(obj, eng):
     return es.count(q='dois.value:"%s"' % (__get_first_doi(obj),))['count'] > 0
 
 
-def store_record(obj, eng):
-    """Stores record in database"""
-    if 'Italiana di Fisica'.lower() in obj.data['abstracts'][0]['source'].lower():
+def _fix_sources(obj):
```
Comment (ammirate) [inline scoap3/modules/workflows/workflows/articles_upload.py]: Maybe better `set_springer_source_if_needed` or something like that? 
Note also the `_`, I would have a test for it with some real cases

File: scoap3/modules/workflows/workflows/articles_upload.py [hunk 12]
```diff
@@ -238,6 +241,27 @@ def build_files_data(obj, eng):
...
+        files = []
+        for document in obj.data.get('documents', ()):
+            if 'xml' in document['key']:
+                filetype = 'xml'
+            elif 'pdfa' in document['key']:
+                filetype = 'pdfa'
+            elif 'pdf' in document['key']:
+                filetype = 'pdf'
+            else:
+                filetype = 'unknown'
```
Comment (ammirate) [inline scoap3/modules/workflows/workflows/articles_upload.py]: What about a more pythonic statement like :point_down: ? 
```python
known_extensions = ('xml', 'pdf', 'pdf')
extension = document['key'] if document['key'] in known_extensions else 'unknown'
```

**Review by ammirate — CHANGES_REQUESTED** [review]
Overall looks good, but some things might be simplified and some functions are missing tests.
Great integration tests BTW!

File: scoap3/modules/robotupload/util.py (hunk 1 above)
Comment (nyirit) [reply scoap3/modules/robotupload/util.py:48]: Yes, I believe we can expect this to be always delivered, and we also use it.

File: scoap3/modules/robotupload/util.py (hunk 2 above)
Comment (nyirit) [reply scoap3/modules/robotupload/util.py:55]: Yes.

File: scoap3/modules/robotupload/util.py (hunk 3 above)
Comment (nyirit) [reply scoap3/modules/robotupload/util.py:62]: We don't use `inspire-crawler` here.
When the delivered xml is parsed using `inspire-dojson`, there won't be an `acquisition_source`, however we expect this in the `article_upload` workflow.

File: scoap3/modules/robotupload/util.py (hunk 4 above)
Comment (nyirit) [reply scoap3/modules/robotupload/util.py:75]: Seems that the `inspire-dojson` uses this field to store the `document_type`. Since in the harvesters we use the `publication_info->material` field as discussed earlier, this should only be present there.
Or if we leave it here, we should remove it later in the `article_upload` workflow.

File: scoap3/modules/workflows/workflows/articles_upload.py (hunk 11 above)
Comment (nyirit) [reply scoap3/modules/workflows/workflows/articles_upload.py]: In fact I'm not sure if we need this. There is only old occurrences in the new repo imported from the legacy repo. The last article like this was delivered in March 2018, so seems like the new repository never used this fix.

File: scoap3/modules/workflows/workflows/articles_upload.py (hunk 12 above)
Comment (nyirit) [reply scoap3/modules/workflows/workflows/articles_upload.py]: The reason for not using this is that I'm checking for a substring, not equation.

**Review by ammirate — APPROVED** [review]
Looks good to me :+1:
